qt/qt_common/map_widget.cpp

```diff
@@ -337,6 +344,64 @@ void MapWidget::initializeGL()
   ASSERT(m_contextFactory == nullptr, ());
   if (!m_screenshotMode)
     m_ratio = devicePixelRatio();
+
+#if defined(OMIM_OS_MAC)
+  m_apiOpenGLES3 = false;
+#elif defined(OMIM_OS_LINUX)
+  {
+    QOpenGLFunctions * funcs = context()->functions();
+    const std::string vendor = QString::fromLatin1(
+      (const char*)funcs->glGetString(GL_VENDOR)).toStdString();
+    const std::string renderer = QString::fromLatin1(
+      (const char*)funcs->glGetString(GL_RENDERER)).toStdString();
+    const std::string version = QString::fromLatin1(
+      (const char*)funcs->glGetString(GL_VERSION)).toStdString();
+    const std::string glslVersion = QString::fromLatin1(
+      (const char*)funcs->glGetString(GL_SHADING_LANGUAGE_VERSION)).toStdString();
+    const std::string extensions = QString::fromLatin1(
+      (const char*)funcs->glGetString(GL_EXTENSIONS)).toStdString();
+
+    LOG(LINFO, ("Vendor:", vendor, "\nRenderer:", renderer,
+                "\nVersion:", version, "\nShading language version:\n", glslVersion,
+                "\nExtensions:", extensions));
+
+    if (context()->isOpenGLES())
+    {
+      LOG(LINFO, ("Context is LibGLES"));
+      // TODO: Fix the ES3 code path with ES3 compatible shader code.
+      m_apiOpenGLES3 = false;
```

Comment:
Does it mean that GLES3 is not working yet on Linux?

Reply:
IIUIC, it means  LibGLES context is currently used for GLES2 only, and for GLES3 - LibGL context is used.

Reply:
@bam80
Not exactly, it's more complicated than that.
The current Qt code path in OMaps (where the relevant code path is ancient), for some inexplicable reason, uses the `m_apiOpenGLES3` boolean to  [switch between two shader codes.](https://github.com/organicmaps/organicmaps/blob/master/qt/qt_common/map_widget.cpp#L203)
1. The first comes without any `#version` directive and therefore is compatible with `OpenGL 2.0` and `OpenGL ES 2.0`
2. The second comes with `#version 150 core` and therefore is only compatible with desktop `OpenGL 3.2` and higher, so no ES.

So if
`m_apiOpenGLES3` is `true` -> Desktop! `OpenGL 3.2` and higher is needed
`m_apiOpenGLES3` is `false` -> Either `OpenGL ES 2.0` or desktop `OpenGL 2.0`

In the past I made an attempt to clean up this ancient mess, as it seemed easy.
The idea is to create a new `ApiVersion` [here](https://github.com/organicmaps/organicmaps/blob/201ab8b0e9bba42631454b2d030290bc0f39280b/drape/drape_global.hpp#L19) so we can finally separate `GLES3` from `GL3`, and use separate shader codes in Qt for `GLES2`, `GLES3` and `GL3`, and forget this `m_apiOpenGLES3` mess.

Now as easy it seemed in theory, in practice have faced some inexplicable shader problems resulting in a completely black map. Nowadays I don't have time to play with that, but I am still confident that this would be the right way forward.

Reply:
@Ferenc- thanks for the details!

1. Why desktop OpenGL 3.2 and higher are needed if `m_apiOpenGLES3 == true`? ES is a subset of the desktop GL, so it should just work, no?
2. Historically, we were targeting only mobile renderers, that's why desktop variants are not present in the enum. Introducing a separate desktop GL branch may be a good idea in the longer term when we have more resources and desktop versions become more popular amongst users. But now it would be much easier to support only mobile ES versions + Vulkan and Metal. What are the cons of this approach in a Linux context?

Reply:
> Why desktop OpenGL 3.2 and higher are needed if m_apiOpenGLES3 == true? ES is a subset of the desktop GL, so it should just work, no?

As I can see, it's actually GLSL version which doesn't work, not GL one.
But GL version implies GLSL one by spec, even if it's artificially.
So we can enforce GL version thus influencing GLSL one implicitly, providing fallback otherwise. That would consists the fix.

Reply:
@Ferenc- `m_apiOpenGLES3 == true` path works perfectly on all androids now. You can force it by typing `?gl` in the search and restarting OM, and return back to vulkan by typing `?vulkan`.

For example, from a recent log:
```
07-06 19:15:26.043 (Thread-17) I/OMcore: drape/support_manager.cpp:35 Init(): Renderer = Mali-G71 | Api = OpenGLES3 | Version = OpenGL ES 3.2 v1.r16p0-01rel0.###other-sha0123456789ABCDEF0###
```